### DISCUSSION TITLE
Manually trigger image transforms generation

### DISCUSSION THREAD

**Discussion by vuolter** [craftcms/cms#11381] [Ideas]
Hello there,

Since we already have in Craft a utility to manually trigger a full re-indexing of assets, I wonder if it would be feasible to have something in the UI or a command in the CLI to force a regeneration of all image transforms.

Generating the transformations on demand, especially if there are many, makes the API unresponsive and, on the other hand, delayed generation means that each request will redirect to the resource once it's ready every time, and this is not great either.

Let me know what you think. Thank you.

**Comment by vuolter** [comment]
Any thoughts?

**Reply by brandonkelly** [reply]
Not yet ported to Craft 4, but [this plugin](https://plugins.craftcms.com/imager-pretransform?cmsConstraint=%5E3.0) should help accomplish what you want if you’re running Craft 3.

**Reply by wsydney76** [reply]
This is now part of the [Imager-X Pro](https://imager-x.spacecat.ninja/usage/generate.html) plugin for both Craft 3 and 4.

If you don't use Imager-X or image sizes are dynamically calculated (e.g. the user selects width/aspect ratio individually) we found the easiest way is to

* set `'generateTransformsBeforePageLoad' => true,` in your `config/general.php`
* call every page once, which forces all transforms to be generated.

That can be done

* by just previewing a page before publishing it...
* if you use a caching plugin by generating (warming) the cache (e.g. craft blitz/cache/generate)
* setting up a custom console command, we use this simple one (which may fail for pages that require request params or a logged in user, but that is acceptable for us): 

```php
/**
     * Creates images transforms by requesting each entry
     *
     * php craft main/assets/create-transforms
     *
     * @throws GuzzleException
     */
    public function actionCreateTransforms(): int
    {

        if ($this->interactive && !$this->confirm('Retrieve each page to create missing image sizes? This may take some time.')) {
            return ExitCode::UNSPECIFIED_ERROR;
        }

        $this->stdout("Retrieving pages" . PHP_EOL);

        $client = Craft::createGuzzleClient();

        // be more specific if there are a ton of pages, e.g. select published by date
        $entries = Entry::find()
            ->uri(':notempty:')
            ->site('*') // uncomment this line if this is a multi site install
            ->unique() // uncomment this line if there are no site specific images in a multisite install
            ->orderBy('id')
            ->all();

        $c = count($entries);
        $i = 0;
        $errors = 0;

        foreach ($entries as $entry) {
            $i++;
            $this->stdout("[{$i}/{$c}] Id: {$entry->id} {$entry->title} ({$entry->site->name})... ");

            try {
                $result = $client->get($entry->getUrl());
                $this->stdout($result->getStatusCode());
            } catch (\Exception $exception) {
                if ($exception->getCode() == 400) {  // or 403? 
                    // Errors can occur if required params are not provided
                    $this->stdout("Error 400, missing params");
                } else {
                    $errors++;
                    $this->stdout("Error {$exception->getMessage()}");
                }
            }

            $this->stdout(PHP_EOL);
        }

        $this->stdout("Done with $errors error(s)" . PHP_EOL);

        return ExitCode::OK;
    }
```

/Mel
